Provide Spire Workload Attestation Support to ztunnel

MikeZappa87 opened the pull request:
# SPIRE Delegated Identity API Integration for ztunnel

## Overview

This document describes the design and implementation of SPIRE integration in ztunnel using the Delegated Identity API. The implementation supports one attestation mode: **PID-based**, each with different security and efficiency trade-offs.

## Background

### Current ztunnel Certificate Management

The existing ztunnel certificate management uses `SecretManager` to cache certificates by `Identity` (SPIFFE ID). When multiple pods share the same service account, they share a single cached certificate, reducing CA calls and memory usage.

### SPIRE Delegated Identity API

SPIRE's Delegated Identity API allows a trusted delegate (ztunnel) to request certificates on behalf of workloads. The API supports two attestation methods:

1. **Selectors**: Identify workloads by Kubernetes namespace + service account (We are not using this as its Spire specific)
2. **PID**: Identify workloads by their process ID for stronger attestation

## Design

### Attestation Modes

In PID mode, each workload is attested individually using its container process ID. This approach:

- Provides stronger security through per-workload attestation
- Each pod receives its own certificate from SPIRE
- Higher SPIRE server load and memory usage
- SPIRE verifies the actual running process, not just Kubernetes metadata

```
┌─────────────────┐     ┌─────────────────┐     ┌─────────────────┐
│  Pod A (PID 1)  │     │  Pod B (PID 2)  │     │  Pod C (PID 3)  │
└────────┬────────┘     └────────┬────────┘     └────────┬────────┘
         │                       │                       │
         ▼                       ▼                       ▼
┌─────────────────┐     ┌─────────────────┐     ┌─────────────────┐
│  Certificate A  │     │  Certificate B  │     │  Certificate C  │
│ (SPIRE call #1) │     │ (SPIRE call #2) │     │ (SPIRE call #3) │
└─────────────────┘     └─────────────────┘     └─────────────────┘
```

### CompositeId Design

#### Motivation

The original `SecretManager` used `Identity` as the cache key. To support PID-based attestation while maintaining backward compatibility with the existing `CaClientTrait` interface, we introduced `CompositeId<RequestKeyEnum>`.

#### Structure

```rust
pub struct CompositeId<RequestKeyEnum> {
    id: Identity,           // The SPIFFE identity (ns/sa)
    key: RequestKeyEnum,    // Distinguishes individual workloads
}

pub enum RequestKeyEnum {
    Identity(Identity),     // For backwards compability
    Workload(WorkloadUid),  // For PID mode: key by workload UID
}
```

#### Trade-offs

This design was chosen to maintain backward compatibility with `CaClientTrait`:

```rust
#[async_trait]
pub trait CaClientTrait: Send + Sync {
    async fn fetch_certificate(
        &self, 
        id: &CompositeId<RequestKeyEnum>
    ) -> Result<tls::WorkloadCertificate, Error>;
}
```

**Benefits:**

- Single interface works for both SPIRE modes and the original CA client
- No breaking changes to existing code paths
- `SecretManager` can track per-workload state when needed

**Consequences:**

- In PID mode, `SecretManager` caches by `CompositeId`, resulting in one cache entry per pod even if they share the same identity
- This is intentional—each workload must be individually attested

### PID Verification Flow

In PID mode, ztunnel performs the following steps:

1. Receive certificate request with `WorkloadUid`
2. Query Container Runtime (CRI) for the container's PID
3. Call SPIRE with the PID for attestation
4. Re-verify PID after SPIRE returns (guards against PID reuse attacks)
5. Return certificate to caller

```rust
async fn get_cert_by_pid(&self, pid: i32, wl_uid: &WorkloadUid) -> Result<...> {
    // 1. Get certificate from SPIRE using PID
    let certs = self.get_cert_from_spire(DelegateAttestationRequest::Pid(pid)).await;
    
    // 2. Re-verify PID hasn't changed (TOCTOU protection)
    if let Some(pid_client) = &self.pid {
        let fetched_pid = pid_client.fetch_pid(wl_uid).await?;
        if fetched_pid.into_i32() != pid {
            return Err(Error::UnableToDeterminePidForWorkload(...));
        }
    }
    
    Ok(certs?)
}
```

## Comparison Summary

| Aspect | Selector Mode | PID Mode |
|--------|---------------|----------|
| **Attestation Granularity** | Per identity (ns/sa) | Per workload (pod) |
| **Certificate Sharing** | Yes—same identity shares cert | No—each pod gets own cert |
| **SPIRE Calls** | 1 per unique identity | 1 per pod |
| **Memory Usage** | Lower | Higher |
| **Security Level** | Standard | Enhanced |
| **Cache Key** | `CompositeId` with `Identity` key | `CompositeId` with `WorkloadUid` key |

## Configuration

```yaml
# Enable SPIRE integration
spire_enabled: true

# Choose attestation mode
spire_mode: "ByPid"  # or "BySelectors"

# SPIRE socket path
spire_socket_path: "/run/spire/sockets/agent.sock"

# Timeout for SPIRE operations  
spire_timeout: "30s"
```

## Future Considerations

1. **Certificate Caching with Per-Pod Attestation**: In PID mode, we should cache and reuse certificates by `Identity` while still attesting every pod individually. This would reduce SPIRE server load and memory usage—multiple pods with the same identity would share one certificate after each pod passes local PID verification. The first pod triggers a SPIRE call; subsequent pods with the same identity only require local PID verification before reusing the cached certificate.

2. **Collaborate with SPIRE/SPIFFE Community**: Work with the SPIRE and SPIFFE community to improve the Delegated Identity API and related interfaces to better support delegated attestation use cases like ztunnel's.

3. Consider a different trait for attested workloads instead of modifying fetch_certificate.

MikeZappa87 committed:
implement spire with delegated identity api

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.
</details>

tjons commented:
@MikeZappa87 hey Mike! This is going to be super helpful... Anything I can do to help move this along? I've got a few commits on SPIRE, do you need help pushing anything on that side forward?

MikeZappa87 commented:
> @MikeZappa87 hey Mike! This is going to be super helpful... Anything I can do to help move this along? I've got a few commits on SPIRE, do you need help pushing anything on that side forward?

The istio community doesn't like the spire specific delegated identity api and want the spiffe broker endpoint api. We are working with the spiffe community to get that moving. Reach out to me on the istio slack, I can add you to the chat.

arndt-s commented on src/identity/manager.rs:
```diff
@@ -576,6 +658,29 @@ impl SecretManager {
+        Ok(Self::new_with_client(client))
+    }
+
+    pub async fn new_with_spire_client_pid<C: 'static + DelegatedIdentityApi>(
```
Is this used?

nddq committed:
docs: add SPIRE CA support section to README

Document that this fork adds SPIRE as an alternative Certificate
Authority for workload identity, and explain the relationship with
Cilium as a temporary sub-project until upstream merge.

Co-authored-by: Bill Mulligan <billmulligan516@gmail.com>
Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

MikeZappa87 committed:
update key structure

nddq committed:
config: bind readiness probe to localhost

Change `readiness_addr` from binding to `0.0.0.0` (all interfaces)
to `127.0.0.1` (localhost only). Since ztunnel runs with
`hostNetwork: true`, binding to all interfaces unnecessarily exposes
the readiness endpoint to the network. The kubelet runs on the same
node and can reach localhost for health probes.

This reduces attack surface for hostNetwork pods.

Update `malicious_calls_inpod` test expectations for port 15021:
captured clients now get a connection reset (Request) and uncaptured
clients get connection refused (Connection) since readiness no longer
listens on the node IP.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

MikeZappa87 committed:
address comments

nddq committed:
ci: add PR tests workflow for formatting and unit tests

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

nddq committed:
ci: add container image build pipeline

Add CI and release workflows for building multi-arch container images.

CI workflow (build-images-ci.yaml):
- Triggers on push to master and PRs
- Builds binaries for amd64/arm64 with both aws-lc and boring TLS modes
- Pushes multi-arch images to dev registry with SHA tags
- FIPS variant uses tag suffix (-fips) on the same repo

Release workflow (build-images-releases.yaml):
- Triggers on version tags (v*)
- Same binary matrix as CI
- Pushes to release registry with version tags
- Creates GitHub Release with auto-generated changelog
- Attaches binaries and container image tarballs as release assets

Supporting files:
- Dockerfile using distroless base with TARGETARCH for multi-arch
- Makefile.docker with local build and multi-arch push targets
- build-release target added to Makefile.core.mk

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

MikeZappa87 committed:
address pr comments

MikeZappa87 committed:
add test for pid change

MikeZappa87 committed:
fix issue with exited container

nddq committed:
ci: fix release workflow artifact download and environment

Filter artifact download to ztunnel-* pattern to avoid downloading
unrelated dockerbuild metadata artifacts. Add release environment to
create-release job so Quay.io secrets are available for skopeo login.

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

MikeZappa87 committed:
fix linter

nddq committed:
Add Unix socket support for XDS address

Enable xds_address to use Unix domain sockets via the unix:// URI
scheme. This allows ztunnel to connect to local control plane
components through Unix sockets instead of only TCP/TLS.

Changes:
- Add unix_socket_path() helper for string-based unix: URI detection
  (hyper's Uri parser does not support the unix scheme)
- Add UdsConnector and UdsGrpcChannel for Unix socket gRPC connections
- Add GrpcChannel enum to support both TLS and Unix socket channels
- Update grpc_channel() factory to route unix: URIs to UDS path
- Update XDS client to detect Unix scheme and skip TLS config
- Validate unix: URIs in config to reject empty socket paths
- Revert CA client to TLS-only (Unix socket scoped to XDS)
- Add unit tests for URI validation, channel factory, and connector

Unix sockets don't require TLS as they are local trusted paths. Auth
headers are still injected for control plane authentication.

Usage:
  export XDS_ADDRESS="unix:///var/run/xds.sock"

Signed-off-by: Quang Nguyen <nguyenquang@microsoft.com>

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.
</details>

MikeZappa87 committed:
wrap both rpcs in timeout

MikeZappa87 commented:
This is going to be replaced with the spiffe broker endpoint api.

yashwanth-l commented:
Should this PR be closed, as **superseded** by [istio/ztunnel#1936](https://github.com/istio/ztunnel/pull/1936) ?
